src/coreclr/vm/debuginfostore.cpp

```diff
@@ -711,33 +783,43 @@ void CompressDebugInfo::EnumMemoryRegions(CLRDataEnumMemoryFlags flags, PTR_BYTE
     }
     CONTRACTL_END;
 
-#ifdef FEATURE_ON_STACK_REPLACEMENT
+    PTR_BYTE pStart = pDebugInfo;
+
     if (hasFlagByte)
     {
         // Check flag byte and skip over any patchpoint info
         BYTE flagByte = *pDebugInfo;
         pDebugInfo++;
 
-        if (flagByte == 1)
+        if ((flagByte & EXTRA_DEBUG_INFO_PATCHPOINT) != 0)
         {
             PTR_PatchpointInfo patchpointInfo = dac_cast<PTR_PatchpointInfo>(pDebugInfo);
             pDebugInfo += patchpointInfo->PatchpointInfoSize();
+            flagByte &= ~EXTRA_DEBUG_INFO_PATCHPOINT;
         }
-        else
+
+        if ((flagByte & EXTRA_DEBUG_INFO_RICH) != 0)
         {
-            _ASSERTE(flagByte == 0);
+            UINT32 iInlineTree = *PTR_UINT32(pDebugInfo);
+            pDebugInfo += 4;
+            UINT32 iRichMappings = *PTR_UINT32(pDebugInfo);
+            pDebugInfo += 4;
+            pDebugInfo += iInlineTree * sizeof(ICorDebugInfo::InlineTreeNode);
+            pDebugInfo += iRichMappings * sizeof(ICorDebugInfo::RichOffsetMapping);
+            flagByte &= ~EXTRA_DEBUG_INFO_RICH;
         }
+
+        _ASSERTE(flagByte == 0);
     }
-#else
-    _ASSERTE(!hasFlagByte);
-#endif
 
     NibbleReader r(pDebugInfo, 12 /* maximum size of compressed 2 UINT32s */);
 
     ULONG cbBounds = r.ReadEncodedU32();
     ULONG cbVars   = r.ReadEncodedU32();
 
-    DacEnumMemoryRegion(dac_cast<TADDR>(pDebugInfo), r.GetNextByteIndex() + cbBounds + cbVars);
+    pDebugInfo += r.GetNextByteIndex() + cbBounds + cbVars;
+
+    DacEnumMemoryRegion(dac_cast<TADDR>(pStart), pDebugInfo - pStart);
```

Comment:
This code also seemed wrong before -- it would not include all the debug information in some cases.